### PR TITLE
feat(compat): update patch for typescript 4.4

### PR DESCRIPTION
**Pull request by merceyz**
**What's the problem this PR addresses?**

- The PnP patch for TypeScript doesn't apply to `typescript@>=4.4.1-rc`
- To quote the test file:
```
The PnP patch returned the packageLocation with a trailing `/` which
caused TypeScript to calculate the `subModuleName` incorrectly[1].
This caused different files where only the first character of their
basename was different to end up with the same packageIdKey[2].
In this test the affected files were
@types/lodash-es/find.d.ts -> @types/lodash-es/ind.d.ts@4.17.4
@types/lodash-es/bind.d.ts -> @types/lodash-es/ind.d.ts@4.17.4

1: https://github.com/microsoft/TypeScript/blob/98866a5a740b487c20046d4ffaa36aa1f202dde9/src/compiler/moduleNameResolver.ts#L20
2: https://github.com/arcanis/TypeScript/blob/36225c32609137d29008e9d3ecf48c6f51456eb5/src/compiler/program.ts#L2625
```
- The references patch doesn't preserve paths when resolving virtuals and assumes everything points to a package root

Fixes https://github.com/yarnpkg/berry/issues/3058
Fixes https://github.com/yarnpkg/berry/issues/3231
Fixes https://github.com/yarnpkg/berry/issues/3117
Fixes https://github.com/yarnpkg/berry/issues/3350
Closes https://github.com/yarnpkg/berry/issues/3275
Closes https://github.com/arcanis/TypeScript/pull/10

**How did you fix it?**

- Rebased https://github.com/microsoft/TypeScript/pull/35206 onto the `4.4.1-rc` release
- Strip the trailing slash (https://github.com/merceyz/TypeScript/commit/98866a5a740b487c20046d4ffaa36aa1f202dde9)
- Preserve the path after the `packageLocation` (https://github.com/merceyz/TypeScript/commit/3f4e1516e0a956e1e0d46d9caf2a1f896fb77aaa)

**Checklist**
- [x] I have read the [Contributing Guide](https://yarnpkg.com/advanced/contributing).
- [x] I have set the packages that need to be released for my changes to be effective.
- [x] I will check that all automated PR checks pass before the PR gets reviewed.

### CODE REVIEW

**Comment by merceyz** [comment]
Failing CI will be fixed by https://github.com/yarnpkg/berry/pull/3315

File: packages/plugin-compat/extra/typescript/gen-typescript-patch.js [hunk 1]
```diff
@@ -267,21 +288,24 @@ async function run({from, to, onto, range}) {
 }
 
 async function validate(version, patchFile) {
```
Comment (merceyz) [inline packages/plugin-compat/extra/typescript/gen-typescript-patch.js:290]: WSL doesn't clean the `/tmp` folder so I had about 10GB of TypeScript there, the changes here should avoid that and improve the performance of repeated runs

**Comment by SimenB** [comment]
Will this be backported to v2? I have a bunch of projects which still supports node 10

**Comment by sushruth** [comment]
This may have broken some base scenarios - https://github.com/sushruth/simple

here is a simple repro repo which imports react and console logs it. TSC finds that erroneous like so - 

image

**Comment by merceyz** [comment]
It didn't, see https://github.com/yarnpkg/berry/issues/3350#issuecomment-906789615

**Comment by sushruth** [comment]
@merceyz still fails. It fails for all packages which have typings included too. Added immer to show this - 
image

**Comment by merceyz** [comment]
Read the rest of the comments in the issue I linked please

**Comment by sushruth** [comment]
@merceyz my bad. Thanks!

**Comment by arcanis** [comment]
> Will this be backported to v2? I have a bunch of projects which still supports node 10

@SimenB yes, I'm in the process of cherry-picking it on 2.4; I'll have to afk, but I'll make sure it gets released tomorrow.

**Comment by SimenB** [comment]
Awesome, thanks! 🎉

**Comment by arcanis** [comment]
Released 2.4.3

**Comment by SimenB** [comment]
Thanks! Can confirm it works 👍

**Comment by mrienstra** [comment]
These changes were released in 3.0.2: [@yarnpkg/cli/3.0.2](https://github.com/yarnpkg/berry/releases/tag/%40yarnpkg%2Fcli%2F3.0.2), [@yarnpkg/pnp/3.0.2](https://github.com/yarnpkg/berry/releases/tag/%40yarnpkg%2Fpnp%2F3.0.2), [@yarnpkg/plugin-pnp/3.0.2](https://github.com/yarnpkg/berry/releases/tag/%40yarnpkg%2Fplugin-pnp%2F3.0.2).
(Old news, just adding for the convenience of others)

**Pull request merged**